internal/core/drand_beacon_public.go

```diff
@@ -55,43 +57,55 @@ func (bp *BeaconProcess) PublicRand(ctx context.Context, in *drand.PublicRandReq
 	beaconResp, err := bp.beacon.Store().Last(ctx)
 	if wanted := in.GetRound(); err == nil && wanted == beaconResp.GetRound()+1 {
 		// we got a request for the next round about to be produced, let's honor it with a callback
-		// we cancel after period since we should never wait so long
-		ctx, cancel := context.WithTimeout(ctx, bp.group.Period)
-		defer cancel()
+		// to make sure we don't miss it because of the aggregation time
+		cctx, cancel := context.WithCancel(ctx)
 		waitlist := make(chan *common.Beacon, 1)
 		rnd := make([]byte, sha256.Size)
 		_, err = rand.Read(rnd)
 		if err != nil {
-			rnd = beaconResp.GetRandomness()[:sha256.Size]
+			_ = copy(rnd, beaconResp.GetRandomness())
 		}
-		fn := func(b *common.Beacon, closed bool) {
-			if closed {
-				close(waitlist)
+		var mu sync.Mutex
+		// the closed bool is only ever called if the same callback is re-added,
+		// this has 0 chances of happening thanks to the rnd in the callback ID.
+		fn := func(b *common.Beacon, _ bool) {
+			mu.Lock()
+			defer mu.Unlock()
+			if cctx.Err() != nil {
 				return
 			}
+			// we can remove our callback as soon as it's executing once
+			bp.beacon.Store().RemoveCallback(addr + hex.EncodeToString(rnd))
+
 			if b.GetRound() == wanted {
 				waitlist <- b
 			}
-			// if we didn't get the right beacon something is off anyway
+			// if we didn't get the right beacon something is off anyway, close
 			close(waitlist)
-			// we can remove our callback once it's executed
-			bp.beacon.Store().RemoveCallback(addr + hex.EncodeToString(rnd))
+			cancel()
 		}
 		bp.beacon.Store().AddCallback(addr+hex.EncodeToString(rnd), fn)
 		select {
 		case <-ctx.Done():
-			return nil, fmt.Errorf("ctx Done in PublicRand: %w", ctx.Err())
+			// make sure to remove callback, noop if already removed
+			bp.beacon.Store().RemoveCallback(addr + hex.EncodeToString(rnd))
```

Comment:
could extract `addr + hex(rnd)` to a var so people don't change one and not the other by accident